File: backend/routes/translate.py
"""Stateless translation helper endpoint - does NOT touch MongoDB. Used by
the (now fully local/offline) Expo app as an optional "online bonus": when
the user creates/edits a custom Feeding/Care/Light&Heat item name while
connected to the internet, the app calls this endpoint to auto-translate the
name to the other language via the Emergent LLM key. If unreachable (no
internet, backend not deployed, etc.) the app silently falls back to using
the same text for both languages - see frontend/localdb/translate.ts."""
from fastapi import APIRouter
from pydantic import BaseModel
from typing import Optional, Literal
from services.translator import translate_item_name

router = APIRouter(prefix="/translate", tags=["translate"])


class TranslateRequest(BaseModel):
    text: str
    source_language: Optional[Literal["en", "da"]] = "en"


@router.post("")
async def translate(payload: TranslateRequest):
    result = await translate_item_name(payload.text, payload.source_language or "en")
    return result

File: frontend/localdb/translate.ts
/**
 * Optional "online bonus" auto-translation for custom Feeding/Care/Light&Heat
 * item names. The app is fully offline/local otherwise - this is the one
 * feature that still calls a (stateless, database-free) backend endpoint,
 * and only when reachable. Never throws: on any failure (offline, backend
 * not deployed, timeout, etc.) it silently falls back to using the same
 * text for both languages, exactly like the old backend's translator.py did
 * when the LLM call failed.
 */
const BASE_URL = `${process.env.EXPO_PUBLIC_BACKEND_URL}/api`;
const TRANSLATE_TIMEOUT_MS = 6000;

export interface Translations {
  da: string;
  en: string;
}

export async function translateItemName(text: string, sourceLanguage: 'en' | 'da'): Promise<Translations> {
  const trimmed = (text || '').trim();
  const fallback: Translations = { da: trimmed, en: trimmed };
  if (!trimmed) return fallback;

  try {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), TRANSLATE_TIMEOUT_MS);
    const res = await fetch(`${BASE_URL}/translate`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ text: trimmed, source_language: sourceLanguage }),
      signal: controller.signal,
    });
    clearTimeout(timeout);
    if (!res.ok) return fallback;
    const data = await res.json();
    return {
      da: (data?.da || trimmed).toString(),
      en: (data?.en || trimmed).toString(),
    };
  } catch {
    // Offline, backend not deployed, timeout, etc. - never break item
    // creation/editing because of this bonus feature.
    return fallback;
  }
}
